### PR TITLE
Battle Tests

### PR DESCRIPTION
**Pull request by mrgriffin**
Documentation at the top of [`test/test_battle.h`](https://github.com/rh-hideout/pokeemerald-expansion/pull/2564/files#diff-cb4f94879153366f4e5f4184a04ff69bd2a792adb03ae2ffceff92d3f29148ec).

Discussed somewhat in Discord already, the aim here is to allow people to write tests for the behavior of things in battle. This is intended to work at the level of moves/items/abilities/etc, I don't think there would be much value in testing lower levels, e.g. I wouldn't bother unit testing the battle script commands themselves. The two uses I can see are:
1. Allows us to do Test Driven Development for new moves/etc. Before implementing, we can write some tests that check all the nuances of the move. Then once they all pass, we know it's correct.
2. Allows us to detect bugs introduced by future development. This was the motivation for the PR, the ability to refactor and be able to check nothing broke.

Uses `mgba-rom-test`, a version of mgba that doesn't have a UI, and so runs much faster. SBird has produced a static binary that in included with this PR. Alternatively you can build `mgba-rom-test` from the mgba repository. You want to set `BUILD_ROM_TEST` to `ON` (either in `CMakeLists.txt` or via something like `cmake -DBUILD_ROM_TEST=ON`).

You can run all tests with `make check` or a subset of tests with `make check TESTS='prefix'` where only tests that start with `prefix` will run.

Notes and limitations:
- Features many macros, some uses of the linker, and the hanging test detection (`Intr_Timer2`) does some fiddly things.
- The implementation calls the test function multiple times, so it's vital that the tests don't mutate any global state. This could be supported, but at the cost of more complexity in the test runner and macros.
- Piggybacks on Recorded Battles, which means a few things are impossible to test including: experience, bag items, wild Pokémon.
- There's no sensible way to test configurations other than the default.
- There's limited support for RNG. I think full support is probably best done with changes across the battle codebase, probably using the linker to allow us to take control of the random number generation. We can run tests a couple of hundred times and see if the proportion of successes is what we'd expect (via `PASSES_RANDOMLY`), but these are very slow.
- Rolls its own (too) basic implementation of `printf` because we can't rely on `mini_printf` being enabled.
- The test runner aims to separate the battle-specific stuff so that people could add other kinds of tests later if they wanted.

### CODE REVIEW

**Comment by mrgriffin** [comment]
Ready for review. Documentation at the top of [`test/test_battle.h`](https://github.com/rh-hideout/pokeemerald-expansion/pull/2564/files#diff-cb4f94879153366f4e5f4184a04ff69bd2a792adb03ae2ffceff92d3f29148ec).

**Comment by SBird1337** [comment]
Did not do a full review yet, and this will probably take some time if you think me reviewing this is a hard requirement. (I'm fine with someone else doing it / you self-approving it)

However I'd like to sort out the `mgba-rom-test` building issue before. I drafted a PR with prebuilt static binaries for windows and linux that should hopefully sort it out for a majority of platforms this has to run on.

File: ld_script_test.txt [hunk 1]
```diff
@@ -0,0 +1,1369 @@
...
+    .debug_str      0 : { *(.debug_str) }
+    .debug_loc      0 : { *(.debug_loc) }
+    .debug_macinfo  0 : { *(.debug_macinfo) }
+
+    /* Discard everything not specifically mentioned above. */
+    /DISCARD/ :
+    {
+        *(*);
+    }
+}
```
Comment (SBird1337) [inline ld_script_test.txt:1369]: Does this have to specify all files in order?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -28,8 +28,14 @@ jobs:
           path: agbcc
           repository: pret/agbcc
 
+      - name: Checkout mgba
+        uses: actions/checkout@v2
+        with:
+          path: mgba
+          repository: mgba-emu/mgba
+
```
Comment (SBird1337) [inline .github/workflows/build.yml:36]: If the static build works we do not need to chceckout mgba.

File: ld_script_test.txt (hunk 1 above)
Comment (mrgriffin) [reply ld_script_test.txt:1369]: Oh, haha, definitely not :)

**Review by SBird1337 — APPROVED** [review]

**Review by SBird1337 — DISMISSED** [review]

**Comment by mrgriffin** [comment]
Don't merge yet: I'm working on a piece of middleware to handle running multiple `mgba-rom-test` processes to speed up the tests.

**Comment by SBird1337** [comment]
I will leave this open until tonight in case anyone has anything to add / review still. If there are no more considerations I will merge it s.t. we can continue with testing upcoming PRs as well as the rest of the currently missing battle mechanics.

**Comment by mrgriffin** [comment]
I have copied the contents of my test-related notepad below:

Notes for testers:
- Should decap any `MESSAGE`s and the corresponding string in `src/battle_message.c`.
- **Must not** mutate global state in the test. And especially not call something like `Random`.

Notes for maintainers:
- Turning off the IRQs, using the timer, overwriting `INTR_VECTOR`, or overwriting `IntrMain_Buffer` will break the abort IRQ. (The latter two could occur on stack overflow).
- The test ROM needs to be under 32MB. Might make sense to have a more explicit linker script if we run out of space.
- Can't use `libisagbprn`/`mini_printf` because the functions will not exist if `NDEBUG` is defined.

Bugs:
- The terrain flag and thunderstorm field weather apply to recorded battles (but not to battle weather, which is the correct behavior).
- Encore should be allowed to pick its target.
- What is Leftovers' `holdEffectParam` for?

Unused:
- `EFFECT_SPEED_UP`
- `EFFECT_SPECIAL_DEFENSE_UP`
- `EFFECT_ACCURACY_UP`
- `EFFECT_EVASION_UP_2`
- `EFFECT_SPECIAL_DEFENSE_DOWN`

Untested (needs configuration support):
- `EFFECT_SPECIAL_ATTACK_UP` (no moves past gen 5)
- `EFFECT_SPEED_DOWN` (no moves past gen 6)
- `EFFECT_EVASION_DOWN` (no moves past gen 6)

Purposefully untested:
- The particular values in `src/data/battle_moves.h`. Perhaps these could be verified against a database from Smogon or whatever.
- "Can't apply a status to an already statused Pokémon" is not tested in every effect that can apply a status. It _should_ be tested in `test/status.c` (but currently isn't).

Battle API improvements:
- `MOVE_ANIMATION(move, battler, [target:])` to streamline `ANIMATION(ANIM_TYPE_MOVE, ...)`?
- `ANIMATION` should be able to inspect the animation parameter (useful for stat animations and maybe others).
- Way to match a switch-in (the switch out animation can be matched, but not the switch in one).
- AI tests. Need a way to set the flags. Probably want a way to have particular actions controlled by the AI. Can patch the action in `BattleTest_CheckBattleRecordActionType`?
- `PASSES_RANDOMLY` should be able to specify which `SCENE`/`THEN` checks in particular are affected by randomness.
- Potentially allow the `MESSAGE` pattern to specify case-insensitivity to support capitalization/decapitalization? (Probably implemented as a preproc flag that applies globally and defaults to being strict)
- Extend `NOT` to support blocks.
- Reintroduce `ONLY_IF` but not using `ONE_OF` for the `TRUE` case so that it can be used with blocks. `NONE_OF` also potentially has strange semantics.
- Potentially an `OPTIONAL` matcher writes to a `bool8` whether or not the `SCENE` command matched?
- Make the stat-changing functions manipulate IVs/EVs to reach the target stat? This interacts better with Mega Evolution, and plays into the concept of realistic tests.
- Support 1v2 and 2v1 battles.
- Support wild battles.
- Support bag items (including Poké Balls in wild battles).

Non-battle API improvements:
- `CYCLES { ... }` that times the number of cycles inside the block. Useful for writing tests of optimizations (the old code would be moved into the test suite and would also serve as an oracle for testing the functionality).

Expansion improvements:
- Create an `ASSERT` macro which calls `Test_ExitWithResult` if `gTestRunnerEnabled` (rather than terminating the code).
- Prevent fainted Pokémon being sent out.
- Instrument `Random()` calls in the battle engine so that `PASSES_RANDOMLY` can use that information instead of having to run the battle 50+ times (see [Discord](https://discord.com/channels/419213663107416084/774393519569502268/1069292392894705736) for a potential implementation/API). This could also be useful for a simulation-based AI.
- Make Z moves available by changing `QueueZMove` into a flag like `RET_MEGA_EVOLUTION`. Then allow `MOVE` to set this flag.

Hydra/runner improvements:
- Set `gTestRunnerHeadless` in Hydra rather than having `make` generate it (requires passing `SKIP_IS_FAIL`).
- Summarize the number of tests run.
- See if we can submit a PR to mgba which adds a pipe that can be read to pull a byte from stdin, and written to push a byte to stdout. Note we'll need a second register to say whether the pipe contains data. We can then use this to configure our tests instead of `patchelf`. It also opens the door to having Hydra manage which tests to run on which workers, rather than the current scheduling algorithm.
- Static analysis of `#ifdef`s (etc) to determine which battle configurations (and combinations thereof) need testing.
- Skip the battle intro to improve performance. Maybe via [SBird's faster battle intros](https://github.com/pret/pokeemerald/compare/master...SBird1337:pokeemerald:battle/simple-intro). It's better to have this be a (non-preproc-controlled) feature rather than a test-only thing.
- Use a non-standard battler layout to catch examples of the code referencing `gBattleMons` by position rather than by looking up the index using `GetBattlerPosition`.
- Cause the test to fail if memory is leaking?

To test:
- Damage variation.
- Poison Point (etc) are prevented by Safeguard.
- Thunder Wave is affected by type immunity, but other `EFFECT_PARALYZE` are not.
- Test for every animation that ensure that it doesn't crash or timeout.

**Review by SBird1337 — APPROVED** [review]

**Pull request merged**